public/app/plugins/panel/timeseries/plugins/ThresholdDragHandle.tsx

```diff
@@ -0,0 +1,83 @@
+import React, { useState } from 'react';
+import { css } from '@emotion/css';
+import { Threshold, GrafanaTheme2 } from '@grafana/data';
+import { useTheme2 } from '@grafana/ui';
+import Draggable, { DraggableBounds } from 'react-draggable';
+
+interface ThresholdDragHandleProps {
+  step: Threshold;
+  y: number;
+  dragBounds: DraggableBounds;
+  mapPositionToValue: (y: number) => number;
+  onChange: (value: number) => void;
+  formatValue: (value: number) => string;
+}
+
+export const ThresholdDragHandle: React.FC<ThresholdDragHandleProps> = ({
+  step,
+  y,
+  dragBounds,
+  mapPositionToValue,
+  formatValue,
+  onChange,
+}) => {
+  const theme = useTheme2();
+  const styles = getStyles(theme, step);
+  const [currentValue, setCurrentValue] = useState(step.value);
+
+  return (
+    <Draggable
+      axis="y"
+      grid={[1, 1]}
+      onStop={(_e, d) => onChange(mapPositionToValue(d.lastY))}
+      onDrag={(_e, d) => {
+        setCurrentValue(mapPositionToValue(d.lastY));
+      }}
+      position={{ x: 0, y }}
+      bounds={dragBounds}
+    >
+      <div className={styles.handle}>
+        <span className={styles.handleText}>{formatValue(currentValue)}</span>
+      </div>
+    </Draggable>
+  );
+};
+
+ThresholdDragHandle.displayName = 'ThresholdDragHandle';
+
+const getStyles = (theme: GrafanaTheme2, step: Threshold) => {
```

Comment:
i'm willing to bet that the drag lag is due to this function regenerating the style hash and setting all of these props on every mousemove ;)

Reply:
Yes, this is POC, no optimization yet. This fn needs to be either memoized or tyles applied inline.

Reply:
fixed now with inline styles and getStyles memoization